### ISSUE TITLE
What's the correct way to handle the browser being able to reach client-only routes directly?

### ISSUE THREAD

**Issue by ariadne-github** [gatsbyjs/gatsby#13443]
## Summary

Referring to this page of the documentation: https://www.gatsbyjs.org/docs/building-apps-with-gatsby/#client-only-routes--user-authentication

> If you’d like site users to be able to visit client routes directly, you’ll need to set up your server to handle those routes appropriately.

This part is not clear, it doesn't explain clearly what the server should do.
I come across [#12576](https://github.com/gatsbyjs/gatsby/issues/12576) and what I understood, at least for simple cases, is that it should redirect to 
the corresponding static file rendered server side, i.e. /app/index.hml for every route like /app/*.

That said, my question is related to the problems with hydration since react 16 that happen whenever you don't design your dom in a specific way.

To have an idea of what I'm referring to, see [#10370](https://github.com/gatsbyjs/gatsby/issues/10370).

To see another case, that is similar to the project I'm working on, take the simple-auth example, remove the main wrapper inside the layout around children, and add a component like a footer after children (in the layout). 
Then go in app.js and add a "default" attribute to the login route component, for example. 
The first thing is a perfeclty plausible layout choice. 
The second one is an example to try to render something on the server that can be visible on the client before React takes in and render the correct rote hit by the client.
If you try to navigate to /app/profile and then reload the page, you will see (inspecting the page) that the dom gets messed up, because react cannot patch the differences between SSR dom and the client rendered ones. The react docs suggest a [two-pass rendering](https://reactjs.org/docs/react-dom.html#hydrate). 

What is the correct approach in this case in your opinion?

I would expect the router to do handle this in a transparent way, but it doesn't happen.

I'm missing some configuration that can prevent this kind of issues?

I tried different approaches:
- Add a wrapper in the layout around children and remove any default route, having SSR render a empty router, like in the simple-auth.
	This seems to work since react is able to patch whta the router inserts on the client since it's inside what's already rendered. 
	That is my understanding, but seems a bit fragile, becaure it relies on and unsupported case of the react patching algorithm.
- Use a flag to conditionally render a static spinner (or whatever you like) or the entire router, then switch the flag on componentDidMount. 
	This is the approach suggested by the React docs: the SSR renders the spinner, then the first client render renders the spinner again, making React happy.
	Then the flag causes a rerender that kicks in the router, that can then render the correct route.
	

In the second approach the code for app.js would look something like:

```

class App extends React.Component {

	constructor(props){
		super(props);
		this.state = {
			isMounted: false
		}
	}
	
	componentDidMount(){
		this.setState({ isMounted: true });
	}

	render(){
		return <Layout>
			{!isMounted ? (
				<MySpinner/>
			) : (
				<Router>
					<Route1 path="/app/route1"/>
					<Route2 path="/app/route2"/>
					<Redirect from="/app" to="/app/route1" default/>
				</Router>
			)}
		</Layout>
	}

}

```
	
This also allows to handle a redirect without messing with the server side rendered html.
I now stay with the last approach which is much more versatile and safe, but I would like to have your opinion.

It would be better if the router could handle this transparently (or with a simple config option), so that it renders always the corresponding server side html, and than automatically triggers a rerender based on the requested route (if it's different).

I opened an issue to Gatsby (and not to @reach/router) because it's Gatsby that has the context to know what's rendered server side, but correct me if I'm wrong.

**Comment by DSchau** [comment]
A lot to unpack here, so let me attempt to do so.

You've done a great job of summing up the current issue. This is a ReactDOM.hydrate issue, not necessarily anything we can fix in Gatsby. That issue, specifically is:

Inability for ReactDOM.hydrate to render an entirely different tree in server vs. client

Generally - I'd say it's sort of an anti-pattern to render an entirely different tree. Contents within, sure--but generally each page should have a similar structure (e.g. wrapped in a Layout component). However--since it seems this is not the case in your application, the proposal you've outlined seems reasonable, specifically the two-pass rendering.

On the client, you'll use `componentDidMount` to designate client-side rendering, and set up a stateful value to render different content. This is totally reasonable and I would recommend that _if you have this use case_ to use this approach.

Let me also respond to a few of your questions:

> It would be better if the router could handle this transparently (or with a simple config option), so that it renders always the corresponding server side html, and than automatically triggers a rerender based on the requested route (if it's different)

I don't think this is a routing concern. This is an app-level concern, and the approach you've outlined seems to be a reasonable solution. Like I said earlier, I'd generally recommend to render the same content on client and server (as much as possible!) but if your use case doesn't match--you can do something like you've done.

I'm going to close this as answered--but please feel free to reply/re-open if we can help further. Thanks for the question!

**Comment by ariadne-github** [comment]
Thank you for your clarification, but I think I didn't explain myself well about the second part.

I'm fully aware that the issue is caused by trying to render two different things on client and server and that's ultimately a React problem.
The point is, that I'm not trying to do this for a specific app requirement, but because I'm following your suggested approach of client-side routes.
In my understanding, the problem arises because the the server renders just one file, while the client maps that file to multiple client routes. So the server renders one page, while any client requesting any other page finds a mismatch at the first render (if it does not implement a workaround like mine).
This is a consequence of the approach suggested in the Gatsby documentation. Obviously, this problem arises only if you try to reach that routes directly.
In the documentation it's not specified that the client routes approach is wrong in that case, but only that it requires proper server side configuration, and I think I made that correctly.

Correct me if I'm wrong, but I'd like to point out one thing: if you try to reach client-only routes directly, even your examples are fundamentally broken.
They only work because they have a trivial layout that React is able to patch, but relying on that is not recommended by the React docs.
That can be shown in the examples I provided in my previous comment.

So, I may rephrase my question as: "Are client-side routes the correct approach when needed to be reached directly, since in any non trivial case, they arise hydration issues?". 
If the response is negative, that's perfectly fine, I could stick with my workaround or make them static.
In that case I'd only ask you if the limitations of client-side routes can be pointed out clearly in the documentation. 
In the other case, maybe I'm missing something.
